File: codewar/Day-117/stopSpinning.js
// https://www.codewars.com/kata/5264d2b162488dc400000001/train/javascript
/**
 *
 * @param {String} str
 */
function spinWords(str) {
  // TODO: do this, and that also
  return str
    .split(" ")
    .map((item) =>
      item.length >= 5 ? item.split("").reverse().join("") : item
    )
    .join(" ");
}
// console.log(spinWords("hey fellow warriors"));

/* ----------------------------- other solution ----------------------------- */
// this one is good

// this is cool
const reverseWord = (word) => [...word].reverse().join("");
const spingWords2 = (string) => string.replace(/\w{5,}/g, reverseWord);

console.log(spingWords2("hello fellow warriors "));

File: codewar/6-kyu/stopSpinning/stopSpinning.go
package main

import (
	"fmt"
	"strings"
)

func reverseWord(word string) string {
	res := []rune(word)
	for i, j := 0, len(word)-1; i < j; i, j = i+1, j-1 {
		res[i], res[j] = res[j], res[i]
	}
	return string(res)
}

func spinWord(str string) string {
	parts := strings.Fields(str)
	res := []string{}

	for _, value := range parts {
		if len(value) >= 5 {
			res = append(res, reverseWord(value))
		} else {
			res = append(res, value)
		}
	}

	return strings.Join(res, " ")
}

func main() {
	fmt.Println(spinWord("Hey fellow warriors"))
}
